File: BasicML/basicml/optim/adamw.py
from .optimizer             import Optimizer
from typing                 import Optional
from basicml.regularization import Regularizer
from basicml.tensor         import Tensor
import numpy as np

class AdamW(Optimizer):
    def __init__(self,
                 parameters  : list[Tensor],
                 lr          : float,
                 regularizer : Optional[Regularizer] = None,
                 beta1       : float                 = 0.9,
                 beta2       : float                 = 0.999,
                 eps         : float                 = 1e-8,
                 weight_decay: float                 = 0.01):
        super().__init__(parameters, lr, regularizer)
        self.beta1         = beta1
        self.beta2         = beta2
        self.eps           = eps
        self.weight_decay  = weight_decay
        self.step_count    = 0
        self.first_moment  = [np.zeros_like(p.data) for p in self.parameters]
        self.second_moment = [np.zeros_like(p.data) for p in self.parameters]

    def step(self):
        self.step_count += 1
        bias_correction1 = 1.0 - self.beta1 ** self.step_count
        bias_correction2 = 1.0 - self.beta2 ** self.step_count
        for i, param in enumerate(self.parameters):
            if param.requires_grad and param.grad is not None:
                grad = param.grad
                if self.regularizer is not None:
                    grad = grad + self.regularizer.grad(param)
                self.first_moment[i]  = self.beta1 * self.first_moment[i] + (1.0 - self.beta1) * grad
                self.second_moment[i] = self.beta2 * self.second_moment[i] + (1.0 - self.beta2) * grad ** 2
                m_hat = self.first_moment[i] / bias_correction1
                v_hat = self.second_moment[i] / bias_correction2
                param.data -= self.lr * (m_hat / (np.sqrt(v_hat) + self.eps)
                                         + self.weight_decay * param.data)

    def zero_grad(self):
        for param in self.parameters:
            param.zero_grad()

File: AdvanceML/src/optim/adamw.cpp
#include "advanceml/optim/adamw.hpp"

#include <cmath>

namespace advanceml {

AdamW::AdamW(std::vector<Tensor> parameters,
             float learning_rate,
             float beta1,
             float beta2,
             float eps,
             float weight_decay)
    : parameters_(std::move(parameters)),
      learning_rate_(learning_rate),
      beta1_(beta1),
      beta2_(beta2),
      eps_(eps),
      weight_decay_(weight_decay) {
    first_moment_.reserve(parameters_.size());
    second_moment_.reserve(parameters_.size());
    for (const Tensor& param : parameters_) {
        first_moment_.emplace_back(param.data().size(), 0.0f);
        second_moment_.emplace_back(param.data().size(), 0.0f);
    }
}

void AdamW::step() {
    ++step_count_;
    const float bias_correction1 = 1.0f - std::pow(beta1_, step_count_);
    const float bias_correction2 = 1.0f - std::pow(beta2_, step_count_);

    for (size_t i = 0; i < parameters_.size(); ++i) {
        Tensor& param = parameters_[i];
        if (!param.has_grad()) {
            continue;
        }
        Tensor grad = param.grad();
        // Raw pointers and loop-local constants, so the loop has no calls or aliasing through members
        // and vectorizes (sqrt included, given -fno-math-errno).
        const float* const g = grad.data().data();
        FloatBuffer& param_values = param.mutable_data();
        float* const values = param_values.data();
        float* const m = first_moment_[i].data();
        float* const v = second_moment_[i].data();
        const size_t n = param_values.size();
        const float beta1 = beta1_;
        const float beta2 = beta2_;
        const float learning_rate = learning_rate_;
        const float eps = eps_;
        const float weight_decay = weight_decay_;
        for (size_t j = 0; j < n; ++j) {
            m[j] = beta1 * m[j] + (1.0f - beta1) * g[j];
            v[j] = beta2 * v[j] + (1.0f - beta2) * g[j] * g[j];
            const float m_hat = m[j] / bias_correction1;
            const float v_hat = v[j] / bias_correction2;
            values[j] -= learning_rate * (m_hat / (std::sqrt(v_hat) + eps) + weight_decay * values[j]);
        }
    }
}

void AdamW::zero_grad() {
    for (Tensor& param : parameters_) {
        param.zero_grad();
    }
}

}  // namespace advanceml
